homeassistant/components/mqtt/alarm_control_panel.py

```diff
@@ -177,38 +173,40 @@ def _setup_from_config(self, config: ConfigType) -> None:
             self._attr_code_format = alarm.CodeFormat.TEXT
         self._attr_code_arm_required = bool(self._config[CONF_CODE_ARM_REQUIRED])
 
+    def _state_message_received(self, msg: ReceiveMessage) -> None:
+        """Run when new MQTT message has been received."""
+        payload = self._value_template(msg.payload)
+        if payload not in (
+            STATE_ALARM_DISARMED,
+            STATE_ALARM_ARMED_HOME,
+            STATE_ALARM_ARMED_AWAY,
+            STATE_ALARM_ARMED_NIGHT,
+            STATE_ALARM_ARMED_VACATION,
+            STATE_ALARM_ARMED_CUSTOM_BYPASS,
+            STATE_ALARM_PENDING,
+            STATE_ALARM_ARMING,
+            STATE_ALARM_DISARMING,
+            STATE_ALARM_TRIGGERED,
+        ):
+            _LOGGER.warning("Received unexpected payload: %s", msg.payload)
+            return
+        self._attr_state = str(payload)
+
     def _prepare_subscribe_topics(self) -> None:
         """(Re)Subscribe to topics."""
 
-        @callback
-        @log_messages(self.hass, self.entity_id)
-        @write_state_on_attr_change(self, {"_attr_state"})
-        def message_received(msg: ReceiveMessage) -> None:
-            """Run when new MQTT message has been received."""
-            payload = self._value_template(msg.payload)
-            if payload not in (
-                STATE_ALARM_DISARMED,
-                STATE_ALARM_ARMED_HOME,
-                STATE_ALARM_ARMED_AWAY,
-                STATE_ALARM_ARMED_NIGHT,
-                STATE_ALARM_ARMED_VACATION,
-                STATE_ALARM_ARMED_CUSTOM_BYPASS,
-                STATE_ALARM_PENDING,
-                STATE_ALARM_ARMING,
-                STATE_ALARM_DISARMING,
-                STATE_ALARM_TRIGGERED,
-            ):
-                _LOGGER.warning("Received unexpected payload: %s", msg.payload)
-                return
-            self._attr_state = str(payload)
-
         self._sub_state = subscription.async_prepare_subscribe_topics(
             self.hass,
```

Comment:
When we finish all the platforms this block will might be the same for all of them and could be broken out into the base class. Need to check if that is the result at the end.